websites/T/Taiko Web/presence.ts

```diff
@@ -0,0 +1,114 @@
+const presence = new Presence({
+    clientId: "858246998561783828"
+  }),
+  slideShow = presence.createSlideshow(),
+  browsingStamp = Math.floor(Date.now() / 1000),
+  script = document.createElement("script");
+
+/*
+  Overriding XMLHttpRequest.open method
+  to know the current request song
+  Doesn't work if song is in cache :[
+*/
+script.innerHTML = `
+  !function(open){
+    XMLHttpRequest.prototype.open = function (method, url) {
+      if (url && url.startsWith("https://taiko.uk/taiko/songs/")) {
+        let songId = parseInt(url.substr(29));
+        let song = assets.songs.find(song => song.id === songId);
+        window.song = {};
+        window.song.title_lang = song.title_lang;
+        window.song.title = song.title;
+        window.song.category = song.category;
+      }
+      open.call(this, method, url);
+    }
+  }(XMLHttpRequest.prototype.open);`;
```

Comment:
![image](https://user-images.githubusercontent.com/9348108/123548202-1beab480-d764-11eb-8daf-67e60d7908ab.png)
and btw you are not allowed to access website in this way. If you want to do it use `getPageletiable` and compare like this.

Reply:
I don't think there's any other way to check which song is playing, the whole website is rendered on a canvas and constant use of `getPageletiable` makes the game lag a lot so it's better to avoid more uses of it.

Reply:
Also I am not making any outside requests, just capturing and reading the url of the requests that the site makes, so I guess that should make it ok?

Reply:
If you will use `getPageletiable` outside `UpdateData` event it will not do any lag since it will only execute once.

Reply:
> Also I am not making any outside requests, just capturing and reading the url of the requests that the site makes, so I guess that should make it ok?

and yeah, well no. you still cannot.

Reply:
![image](https://user-images.githubusercontent.com/9348108/123548674-0b3b3e00-d766-11eb-8886-ecb651f36c16.png)

Reply:
yea just put it outside of updatedata and run it every x seconds. Updatedata runs every 100ms or so, thats why it's so laggy.

Reply:
We need to add some option to change interval of it in v3 because presence doesn't need to be updated every 100ms in every situation.

Reply:
[This guy has added a script too](https://github.com/PreMiD/Presences/blob/00c43cf1113b289782d757f77be3238471d83e26/websites/T/Town%20of%20Salem/presence.ts)

Reply:
It's presence from last year where I wasn't reviewer yet. There are new rules and I strictly follow them.

Reply:
And that's game related thing, because from what I see owner even added divs for PreMiD specifically to get this presence working so why shouldn't we allow it?

Reply:
I see your pov, but here are my reasons why this part of the code should stay as it is:

- Showing which song is playing is an important feature for the presence
- There is no other way to tell which song is playing (I have tried and failed, there's no simple variable holding the current song in `window`, and the website is rendered on a canvas so no help from DOM as well)
- The added script isn't malicious and adds no significant overhead

If you don't want to create an exception that's okay.

Reply:
Well I did found. Websocket gives you songID that you can later get from their API even difficulty.
![image](https://user-images.githubusercontent.com/9348108/123551291-6fafca80-d771-11eb-92f0-227f94902c86.png)

Reply:
Wait, does it work for single player as well? It was not in my browser

Reply:
From what I see, yes.

Reply:
`p2.socket.onmessage = e => console.log(e.data);`

Did you do this as well? The id returned isn't song id though

Reply:
I am pretty sure this is the id values for other players online, there's a net play option for players to play together.

Reply:
Lemme see later when I will have time, because there must be a way :)

Reply:
@ririxidev anything on this?

Reply:
I'm busy

Reply:
I will try to do it tomorrow before me going to vacations but I don't really promise. If I will not do that, presence will be allowed to be in store for now in its state.

Reply:
Did some looking around on the page. It seems like when you start playing a game, a lot of useful information is put in the `debugObj.controller` variable (exposed to window), including the song data (`.selectedSong`)

Reply:
I guess this could work

Reply:
This did not work for multiplayer, so I gave up on this idea.